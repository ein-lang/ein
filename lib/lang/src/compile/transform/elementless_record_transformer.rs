use crate::ast::*;
use crate::types::Type;

pub struct ElementlessRecordTransformer {}

impl ElementlessRecordTransformer {
    pub fn new() -> Self {
        Self {}
    }

    pub fn transform(&mut self, module: &Module) -> Module {
        Module::new(
            module.path().clone(),
            module.export().clone(),
            module.imports().to_vec(),
            module.foreign_declarations().to_vec(),
            module.type_definitions().to_vec(),
            module
                .definitions()
                .iter()
                .cloned()
                .chain(
                    module
                        .type_definitions()
                        .iter()
                        .map(|type_definition| {
                            if let Type::Record(record_type) = type_definition.type_() {
                                if record_type.elements().is_empty() {
                                    vec![VariableDefinition::new(
                                        type_definition.name(),
                                        RecordConstruction::new(
                                            record_type.clone(),
                                            Default::default(),
                                            record_type.source_information().clone(),
                                        ),
                                        record_type.clone(),
                                        record_type.source_information().clone(),
                                    )
                                    .into()]
                                } else {
                                    vec![]
                                }
                            } else {
                                vec![]
                            }
                        })
                        .flatten(),
                )
                .collect(),
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::debug::*;
    use crate::types;
    use pretty_assertions::assert_eq;

    #[test]
    fn transform_elementless_record_type_definitions() {
        let record_type = types::Record::new("Foo", Default::default(), SourceInformation::dummy());

        assert_eq!(
            ElementlessRecordTransformer::new().transform(
                &Module::from_definitions_and_type_definitions(
                    vec![TypeDefinition::new("Foo", record_type.clone())],
                    vec![]
                )
            ),
            Module::from_definitions_and_type_definitions(
                vec![TypeDefinition::new("Foo", record_type.clone())],
                vec![VariableDefinition::new(
                    "Foo",
                    RecordConstruction::new(
                        record_type.clone(),
                        Default::default(),
                        SourceInformation::dummy(),
                    ),
                    record_type,
                    SourceInformation::dummy(),
                )
                .into()]
            )
        );
    }
}
