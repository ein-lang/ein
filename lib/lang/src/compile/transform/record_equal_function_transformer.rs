use super::super::error::CompileError;
use super::super::type_comparability_checker::TypeComparabilityChecker;
use super::utilities;
use crate::ast::*;
use crate::types::{self, Type};
use std::sync::Arc;

pub struct RecordEqualFunctionTransformer {
    type_comparability_checker: Arc<TypeComparabilityChecker>,
}

impl RecordEqualFunctionTransformer {
    pub fn new(type_comparability_checker: Arc<TypeComparabilityChecker>) -> Self {
        Self {
            type_comparability_checker,
        }
    }

    pub fn transform(&mut self, module: &Module) -> Result<Module, CompileError> {
        let mut equal_function_definitions = vec![];

        for type_definition in module.type_definitions() {
            if let Type::Record(record_type) = type_definition.type_() {
                if self
                    .type_comparability_checker
                    .check(type_definition.type_())?
                {
                    equal_function_definitions.push(self.create_record_equal_function(record_type));
                }
            }
        }

        Ok(Module::new(
            module.path().clone(),
            module.export().clone(),
            module.export_foreign().clone(),
            module.imports().to_vec(),
            module.import_foreigns().to_vec(),
            module.type_definitions().to_vec(),
            module
                .definitions()
                .iter()
                .cloned()
                .chain(equal_function_definitions.into_iter().map(Definition::from))
                .collect(),
        ))
    }

    fn create_record_equal_function(&mut self, record_type: &types::Record) -> FunctionDefinition {
        let source_information = record_type.source_information();
        let mut expression: Expression = Boolean::new(true, source_information.clone()).into();

        for (key, element_type) in record_type.elements() {
            expression = If::new(
                expression,
                EqualityOperation::with_type(
                    element_type.clone(),
                    EqualityOperator::Equal,
                    RecordElementOperation::new(
                        record_type.clone(),
                        key,
                        Variable::new("lhs", source_information.clone()),
                        "$element",
                        Variable::new("$element", source_information.clone()),
                        source_information.clone(),
                    ),
                    RecordElementOperation::new(
                        record_type.clone(),
                        key,
                        Variable::new("rhs", source_information.clone()),
                        "$element",
                        Variable::new("$element", source_information.clone()),
                        source_information.clone(),
                    ),
                    source_information.clone(),
                ),
                Boolean::new(false, source_information.clone()),
                source_information.clone(),
            )
            .into();
        }

        FunctionDefinition::new(
            utilities::get_record_equal_function_name(record_type),
            vec!["lhs".into(), "rhs".into()],
            expression,
            types::Function::new(
                record_type.clone(),
                types::Function::new(
                    record_type.clone(),
                    types::Boolean::new(source_information.clone()),
                    source_information.clone(),
                ),
                source_information.clone(),
            ),
            source_information.clone(),
        )
    }
}
