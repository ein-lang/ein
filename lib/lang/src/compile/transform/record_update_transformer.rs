use super::super::{
    error::CompileError, name_generator::NameGenerator,
    reference_type_resolver::ReferenceTypeResolver,
};
use crate::ast::*;

pub struct RecordUpdateTransformer {
    name_generator: NameGenerator,
}

impl RecordUpdateTransformer {
    pub fn new() -> Self {
        Self {
            name_generator: NameGenerator::new("record_update_argument_"),
        }
    }

    pub fn transform(&mut self, module: &Module) -> Result<Module, CompileError> {
        let reference_type_resolver = ReferenceTypeResolver::new(module);

        module.transform_expressions(&mut |expression| -> Result<Expression, CompileError> {
            if let Expression::RecordUpdate(record_update) = expression {
                let record_type = reference_type_resolver
                    .resolve_to_record(record_update.type_())?
                    .unwrap();
                let source_information = record_update.source_information();
                let name = self.name_generator.generate();

                Ok(Let::new(
                    vec![VariableDefinition::new(
                        &name,
                        record_update.argument().clone(),
                        record_update.type_().clone(),
                        source_information.clone(),
                    )
                    .into()],
                    RecordConstruction::new(
                        record_update.type_().clone(),
                        record_type
                            .elements()
                            .iter()
                            .map(|element| {
                                (
                                    element.name().into(),
                                    RecordElementOperation::new(
                                        record_update.type_().clone(),
                                        element.name(),
                                        Variable::new(&name, source_information.clone()),
                                        source_information.clone(),
                                    )
                                    .into(),
                                )
                            })
                            .chain(record_update.elements().clone())
                            .collect(),
                        source_information.clone(),
                    ),
                    source_information.clone(),
                )
                .into())
            } else {
                Ok(expression.clone())
            }
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{debug::*, types};
    use pretty_assertions::assert_eq;

    #[test]
    fn transform_record_update() {
        let record_type = types::Record::new(
            "Foo",
            vec![
                types::RecordElement::new("foo", types::None::new(SourceInformation::dummy())),
                types::RecordElement::new("bar", types::None::new(SourceInformation::dummy())),
            ],
            SourceInformation::dummy(),
        );
        let reference_type = types::Reference::new("Foo", SourceInformation::dummy());

        assert_eq!(
            RecordUpdateTransformer::new().transform(
                &Module::from_definitions_and_type_definitions(
                    vec![TypeDefinition::new("Foo", record_type.clone())],
                    vec![VariableDefinition::new(
                        "x",
                        RecordUpdate::new(
                            reference_type.clone(),
                            Variable::new("foo", SourceInformation::dummy()),
                            vec![("bar".into(), None::new(SourceInformation::dummy()).into())]
                                .into_iter()
                                .collect(),
                            SourceInformation::dummy()
                        ),
                        reference_type.clone(),
                        SourceInformation::dummy(),
                    )
                    .into()]
                )
            ),
            Ok(Module::from_definitions_and_type_definitions(
                vec![TypeDefinition::new("Foo", record_type)],
                vec![VariableDefinition::new(
                    "x",
                    Let::new(
                        vec![VariableDefinition::new(
                            "record_update_argument_0",
                            Variable::new("foo", SourceInformation::dummy()),
                            reference_type.clone(),
                            SourceInformation::dummy(),
                        )
                        .into()],
                        RecordConstruction::new(
                            reference_type.clone(),
                            vec![
                                (
                                    "foo".into(),
                                    RecordElementOperation::new(
                                        reference_type.clone(),
                                        "foo",
                                        Variable::new(
                                            "record_update_argument_0",
                                            SourceInformation::dummy()
                                        ),
                                        SourceInformation::dummy(),
                                    )
                                    .into()
                                ),
                                ("bar".into(), None::new(SourceInformation::dummy()).into())
                            ]
                            .into_iter()
                            .collect(),
                            SourceInformation::dummy()
                        ),
                        SourceInformation::dummy(),
                    ),
                    reference_type,
                    SourceInformation::dummy(),
                )
                .into()]
            ))
        );
    }
}
