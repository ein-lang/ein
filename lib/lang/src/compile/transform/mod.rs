mod boolean_operation_transformer;
mod elementless_record_transformer;
mod equal_operation_transformer;
mod function_type_coercion_transformer;
mod list_case_transformer;
mod list_literal_transformer;
mod not_equal_operation_transformer;
mod record_element_function_transformer;
mod record_equal_function_transformer;
mod record_update_transformer;
mod type_coercion_transformer;
mod typed_meta_transformer;
mod utilities;

use super::compile_configuration::CompileConfiguration;
use super::error::CompileError;
use super::expression_type_extractor::ExpressionTypeExtractor;
use super::last_result_type_calculator::LastResultTypeCalculator;
use super::module_environment_creator::ModuleEnvironmentCreator;
use super::reference_type_resolver::ReferenceTypeResolver;
use super::type_canonicalizer::TypeCanonicalizer;
use super::type_comparability_checker::TypeComparabilityChecker;
use super::type_equality_checker::TypeEqualityChecker;
use crate::ast::*;
pub use boolean_operation_transformer::BooleanOperationTransformer;
use elementless_record_transformer::ElementlessRecordTransformer;
pub use equal_operation_transformer::EqualOperationTransformer;
pub use function_type_coercion_transformer::FunctionTypeCoercionTransformer;
pub use list_case_transformer::ListCaseTransformer;
pub use list_literal_transformer::ListLiteralTransformer;
pub use not_equal_operation_transformer::NotEqualOperationTransformer;
use record_element_function_transformer::RecordElementFunctionTransformer;
use record_equal_function_transformer::RecordEqualFunctionTransformer;
use record_update_transformer::RecordUpdateTransformer;
use std::sync::Arc;
use type_coercion_transformer::TypeCoercionTransformer;
use typed_meta_transformer::TypedMetaTransformer;

pub fn transform_before_name_qualification(module: &Module) -> Result<Module, CompileError> {
    let reference_type_resolver = ReferenceTypeResolver::new(module);
    let type_comparability_checker = TypeComparabilityChecker::new(reference_type_resolver);

    let module = ElementlessRecordTransformer::new().transform(&module);
    let module = RecordElementFunctionTransformer::new().transform(&module);
    let module =
        RecordEqualFunctionTransformer::new(type_comparability_checker).transform(&module)?;

    Ok(module)
}

pub fn transform_without_types(module: &Module) -> Result<Module, CompileError> {
    RecordUpdateTransformer::new().transform(module)
}

pub fn transform_with_types(
    module: &Module,
    compile_configuration: Arc<CompileConfiguration>,
) -> Result<Module, CompileError> {
    let reference_type_resolver = ReferenceTypeResolver::new(module);
    let type_equality_checker = TypeEqualityChecker::new(reference_type_resolver.clone());
    let type_canonicalizer = TypeCanonicalizer::new(
        reference_type_resolver.clone(),
        type_equality_checker.clone(),
    );
    let expression_type_extractor = ExpressionTypeExtractor::new(
        reference_type_resolver.clone(),
        type_equality_checker.clone(),
        type_canonicalizer.clone(),
        compile_configuration.error_type_configuration.clone(),
    );
    let last_result_type_calculator =
        LastResultTypeCalculator::new(reference_type_resolver.clone());
    let module_environment_creator =
        ModuleEnvironmentCreator::new(compile_configuration.builtin_configuration.clone());

    let mut type_coercion_transformer = TypedMetaTransformer::new(
        TypeCoercionTransformer::new(
            reference_type_resolver.clone(),
            type_equality_checker,
            expression_type_extractor,
            type_canonicalizer,
            last_result_type_calculator,
        ),
        module_environment_creator,
        reference_type_resolver,
    );

    type_coercion_transformer.transform(&module)
}

#[cfg(test)]
mod tests {
    use super::super::compile_configuration::COMPILE_CONFIGURATION;
    use super::*;
    use crate::debug::SourceInformation;
    use crate::types;
    use insta::assert_debug_snapshot;

    fn transform_with_types(module: &Module) -> Result<Module, CompileError> {
        super::transform_with_types(module, COMPILE_CONFIGURATION.clone())
    }

    mod type_coercion {
        use super::*;
        use pretty_assertions::assert_eq;

        #[test]
        fn transform_function_definition() {
            let union_type = types::Union::new(
                vec![
                    types::Number::new(SourceInformation::dummy()).into(),
                    types::None::new(SourceInformation::dummy()).into(),
                ],
                SourceInformation::dummy(),
            );

            let create_module = |body: Expression| {
                Module::from_definitions(vec![
                    FunctionDefinition::new(
                        "f",
                        vec!["x".into()],
                        Number::new(42.0, SourceInformation::dummy()),
                        types::Function::new(
                            union_type.clone(),
                            types::Number::new(SourceInformation::dummy()),
                            SourceInformation::dummy(),
                        ),
                        SourceInformation::dummy(),
                    )
                    .into(),
                    FunctionDefinition::new(
                        "g",
                        vec!["x".into()],
                        body,
                        types::Function::new(
                            types::Number::new(SourceInformation::dummy()),
                            union_type.clone(),
                            SourceInformation::dummy(),
                        ),
                        SourceInformation::dummy(),
                    )
                    .into(),
                ])
            };

            assert_eq!(
                transform_with_types(&create_module(
                    Application::new(
                        Variable::new("f", SourceInformation::dummy()),
                        Number::new(42.0, SourceInformation::dummy()),
                        SourceInformation::dummy(),
                    )
                    .into()
                )),
                Ok(create_module(
                    TypeCoercion::new(
                        Application::new(
                            Variable::new("f", SourceInformation::dummy()),
                            TypeCoercion::new(
                                Number::new(42.0, SourceInformation::dummy()),
                                types::Number::new(SourceInformation::dummy()),
                                union_type.clone(),
                                SourceInformation::dummy(),
                            ),
                            SourceInformation::dummy(),
                        ),
                        types::Number::new(SourceInformation::dummy()),
                        union_type.clone(),
                        SourceInformation::dummy(),
                    )
                    .into()
                ))
            );
        }

        #[test]
        fn transform_variable_definition() {
            let union_type = types::Union::new(
                vec![
                    types::Number::new(SourceInformation::dummy()).into(),
                    types::None::new(SourceInformation::dummy()).into(),
                ],
                SourceInformation::dummy(),
            );

            let create_module = |expression: Expression| {
                Module::from_definitions(vec![VariableDefinition::new(
                    "x",
                    expression,
                    union_type.clone(),
                    SourceInformation::dummy(),
                )
                .into()])
            };

            assert_eq!(
                transform_with_types(&create_module(
                    Number::new(42.0, SourceInformation::dummy()).into()
                )),
                Ok(create_module(
                    TypeCoercion::new(
                        Number::new(42.0, SourceInformation::dummy()),
                        types::Number::new(SourceInformation::dummy()),
                        union_type.clone(),
                        SourceInformation::dummy(),
                    )
                    .into()
                ))
            );
        }

        #[test]
        fn transform_application() {
            let union_type = types::Union::new(
                vec![
                    types::Number::new(SourceInformation::dummy()).into(),
                    types::None::new(SourceInformation::dummy()).into(),
                ],
                SourceInformation::dummy(),
            );

            let create_module = |argument: Expression| {
                Module::from_definitions(vec![
                    FunctionDefinition::new(
                        "f",
                        vec!["x".into()],
                        Number::new(42.0, SourceInformation::dummy()),
                        types::Function::new(
                            union_type.clone(),
                            types::Number::new(SourceInformation::dummy()),
                            SourceInformation::dummy(),
                        ),
                        SourceInformation::dummy(),
                    )
                    .into(),
                    VariableDefinition::new(
                        "x",
                        Application::new(
                            Variable::new("f", SourceInformation::dummy()),
                            argument,
                            SourceInformation::dummy(),
                        ),
                        types::Number::new(SourceInformation::dummy()),
                        SourceInformation::dummy(),
                    )
                    .into(),
                ])
            };

            assert_eq!(
                transform_with_types(&create_module(
                    Number::new(42.0, SourceInformation::dummy()).into()
                )),
                Ok(create_module(
                    TypeCoercion::new(
                        Number::new(42.0, SourceInformation::dummy()),
                        types::Number::new(SourceInformation::dummy()),
                        union_type.clone(),
                        SourceInformation::dummy(),
                    )
                    .into()
                ))
            );
        }

        #[test]
        fn transform_let_value_expression() {
            let union_type = types::Union::new(
                vec![
                    types::Number::new(SourceInformation::dummy()).into(),
                    types::None::new(SourceInformation::dummy()).into(),
                ],
                SourceInformation::dummy(),
            );

            let create_module = |expression: Expression| {
                Module::from_definitions(vec![VariableDefinition::new(
                    "x",
                    Let::new(
                        vec![VariableDefinition::new(
                            "y",
                            expression,
                            union_type.clone(),
                            SourceInformation::dummy(),
                        )],
                        Number::new(42.0, SourceInformation::dummy()),
                        SourceInformation::dummy(),
                    ),
                    types::Number::new(SourceInformation::dummy()),
                    SourceInformation::dummy(),
                )
                .into()])
            };

            assert_eq!(
                transform_with_types(&create_module(
                    Number::new(42.0, SourceInformation::dummy()).into()
                )),
                Ok(create_module(
                    TypeCoercion::new(
                        Number::new(42.0, SourceInformation::dummy()),
                        types::Number::new(SourceInformation::dummy()),
                        union_type.clone(),
                        SourceInformation::dummy(),
                    )
                    .into()
                ))
            );
        }

        #[test]
        fn transform_let_function_expression() {
            let union_type = types::Union::new(
                vec![
                    types::Number::new(SourceInformation::dummy()).into(),
                    types::None::new(SourceInformation::dummy()).into(),
                ],
                SourceInformation::dummy(),
            );

            let create_module = |body: Expression| {
                Module::from_definitions(vec![
                    FunctionDefinition::new(
                        "f",
                        vec!["x".into()],
                        Number::new(42.0, SourceInformation::dummy()),
                        types::Function::new(
                            union_type.clone(),
                            types::Number::new(SourceInformation::dummy()),
                            SourceInformation::dummy(),
                        ),
                        SourceInformation::dummy(),
                    )
                    .into(),
                    VariableDefinition::new(
                        "x",
                        LetRecursive::new(
                            vec![FunctionDefinition::new(
                                "g",
                                vec!["x".into()],
                                body,
                                types::Function::new(
                                    types::Number::new(SourceInformation::dummy()),
                                    union_type.clone(),
                                    SourceInformation::dummy(),
                                ),
                                SourceInformation::dummy(),
                            )],
                            Number::new(42.0, SourceInformation::dummy()),
                            SourceInformation::dummy(),
                        ),
                        types::Number::new(SourceInformation::dummy()),
                        SourceInformation::dummy(),
                    )
                    .into(),
                ])
            };

            assert_eq!(
                transform_with_types(&create_module(
                    Application::new(
                        Variable::new("f", SourceInformation::dummy()),
                        Number::new(42.0, SourceInformation::dummy()),
                        SourceInformation::dummy(),
                    )
                    .into()
                )),
                Ok(create_module(
                    TypeCoercion::new(
                        Application::new(
                            Variable::new("f", SourceInformation::dummy()),
                            TypeCoercion::new(
                                Number::new(42.0, SourceInformation::dummy()),
                                types::Number::new(SourceInformation::dummy()),
                                union_type.clone(),
                                SourceInformation::dummy(),
                            ),
                            SourceInformation::dummy(),
                        ),
                        types::Number::new(SourceInformation::dummy()),
                        union_type.clone(),
                        SourceInformation::dummy(),
                    )
                    .into()
                ))
            );
        }

        #[test]
        fn transform_record_construction() {
            let union_type = types::Union::new(
                vec![
                    types::Number::new(SourceInformation::dummy()).into(),
                    types::None::new(SourceInformation::dummy()).into(),
                ],
                SourceInformation::dummy(),
            );
            let reference_type = types::Reference::new("Foo", SourceInformation::dummy());

            assert_debug_snapshot!(transform_with_types(
                &Module::from_definitions_and_type_definitions(
                    vec![TypeDefinition::new(
                        "Foo",
                        types::Record::new(
                            "Foo",
                            vec![("foo".into(), union_type.into())]
                                .into_iter()
                                .collect(),
                            SourceInformation::dummy(),
                        )
                    )],
                    vec![VariableDefinition::new(
                        "x",
                        RecordConstruction::new(
                            reference_type.clone(),
                            vec![(
                                "foo".into(),
                                Number::new(42.0, SourceInformation::dummy()).into()
                            )]
                            .into_iter()
                            .collect(),
                            SourceInformation::dummy(),
                        ),
                        reference_type,
                        SourceInformation::dummy(),
                    )
                    .into()],
                )
            ));
        }

        #[test]
        fn transform_union() {
            let lower_union_type = types::Union::new(
                vec![
                    types::Number::new(SourceInformation::dummy()).into(),
                    types::None::new(SourceInformation::dummy()).into(),
                ],
                SourceInformation::dummy(),
            );
            let upper_union_type = types::Union::new(
                vec![
                    types::Boolean::new(SourceInformation::dummy()).into(),
                    types::Number::new(SourceInformation::dummy()).into(),
                    types::None::new(SourceInformation::dummy()).into(),
                ],
                SourceInformation::dummy(),
            );

            let create_module = |expression1: Expression, expression2: Expression| {
                Module::from_definitions(vec![
                    VariableDefinition::new(
                        "x",
                        expression1,
                        lower_union_type.clone(),
                        SourceInformation::dummy(),
                    )
                    .into(),
                    VariableDefinition::new(
                        "y",
                        expression2,
                        upper_union_type.clone(),
                        SourceInformation::dummy(),
                    )
                    .into(),
                ])
            };

            assert_eq!(
                transform_with_types(&create_module(
                    Number::new(42.0, SourceInformation::dummy()).into(),
                    Variable::new("x", SourceInformation::dummy()).into()
                )),
                Ok(create_module(
                    TypeCoercion::new(
                        Number::new(42.0, SourceInformation::dummy()),
                        types::Number::new(SourceInformation::dummy()),
                        lower_union_type.clone(),
                        SourceInformation::dummy(),
                    )
                    .into(),
                    TypeCoercion::new(
                        Variable::new("x", SourceInformation::dummy()),
                        lower_union_type.clone(),
                        upper_union_type.clone(),
                        SourceInformation::dummy(),
                    )
                    .into()
                ))
            );
        }

        #[test]
        fn transform_any() {
            let create_module = |expression: Expression| {
                Module::from_definitions(vec![VariableDefinition::new(
                    "x",
                    expression,
                    types::Any::new(SourceInformation::dummy()),
                    SourceInformation::dummy(),
                )
                .into()])
            };

            assert_eq!(
                transform_with_types(&create_module(
                    Number::new(42.0, SourceInformation::dummy()).into()
                )),
                Ok(create_module(
                    TypeCoercion::new(
                        Number::new(42.0, SourceInformation::dummy()),
                        types::Number::new(SourceInformation::dummy()),
                        types::Any::new(SourceInformation::dummy()),
                        SourceInformation::dummy(),
                    )
                    .into()
                ))
            );
        }
    }

    #[test]
    fn transform_case_expression() {
        let argument_union_type = types::Union::new(
            vec![
                types::Boolean::new(SourceInformation::dummy()).into(),
                types::None::new(SourceInformation::dummy()).into(),
            ],
            SourceInformation::dummy(),
        );
        let result_union_type = types::Union::new(
            vec![
                types::Number::new(SourceInformation::dummy()).into(),
                types::None::new(SourceInformation::dummy()).into(),
            ],
            SourceInformation::dummy(),
        );

        assert_debug_snapshot!(transform_with_types(&Module::from_definitions(vec![
            VariableDefinition::new(
                "x",
                Case::with_type(
                    argument_union_type,
                    "foo",
                    Boolean::new(false, SourceInformation::dummy()),
                    vec![
                        Alternative::new(
                            types::Boolean::new(SourceInformation::dummy()),
                            Number::new(42.0, SourceInformation::dummy()),
                        ),
                        Alternative::new(
                            types::None::new(SourceInformation::dummy()),
                            None::new(SourceInformation::dummy()),
                        ),
                    ],
                    SourceInformation::dummy(),
                ),
                result_union_type,
                SourceInformation::dummy(),
            )
            .into()
        ])));
    }
}
