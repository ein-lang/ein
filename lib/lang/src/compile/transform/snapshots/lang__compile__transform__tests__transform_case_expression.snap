---
source: lib/lang/src/compile/transform/mod.rs
expression: "transform_with_types(&Module::from_definitions(vec![VariableDefinition ::\n                                                    new(\"x\", Case ::\n                                                        with_type(argument_union_type,\n                                                                  \"foo\",\n                                                                  Boolean ::\n                                                                  new(false,\n                                                                      SourceInformation\n                                                                      ::\n                                                                      dummy()),\n                                                                  vec !\n                                                                  [Alternative\n                                                                   ::\n                                                                   new(types\n                                                                       ::\n                                                                       Boolean\n                                                                       ::\n                                                                       new(SourceInformation\n                                                                           ::\n                                                                           dummy()),\n                                                                       Number\n                                                                       ::\n                                                                       new(42.0,\n                                                                           SourceInformation\n                                                                           ::\n                                                                           dummy()),),\n                                                                   Alternative\n                                                                   ::\n                                                                   new(types\n                                                                       :: None\n                                                                       ::\n                                                                       new(SourceInformation\n                                                                           ::\n                                                                           dummy()),\n                                                                       None ::\n                                                                       new(SourceInformation\n                                                                           ::\n                                                                           dummy()),),],\n                                                                  SourceInformation\n                                                                  ::\n                                                                  dummy(),),\n                                                        result_union_type,\n                                                        SourceInformation ::\n                                                        dummy(),) . into()]))"
---
Ok(
    Module {
        path: ModulePath {
            package: Package {
                name: "",
                version: "",
            },
            components: [],
        },
        type_definitions: [],
        definitions: [
            VariableDefinition(
                VariableDefinition {
                    name: "x",
                    body: Case(
                        Case {
                            type_: Union(
                                Union {
                                    types: {
                                        Boolean(
                                            Boolean {
                                                source_information: SourceInformation {
                                                    source_name: "",
                                                    location: Location {
                                                        line_number: 0,
                                                        column_number: 0,
                                                    },
                                                    line: "",
                                                },
                                            },
                                        ),
                                        None(
                                            None {
                                                source_information: SourceInformation {
                                                    source_name: "",
                                                    location: Location {
                                                        line_number: 0,
                                                        column_number: 0,
                                                    },
                                                    line: "",
                                                },
                                            },
                                        ),
                                    },
                                    source_information: SourceInformation {
                                        source_name: "",
                                        location: Location {
                                            line_number: 0,
                                            column_number: 0,
                                        },
                                        line: "",
                                    },
                                },
                            ),
                            name: "foo",
                            argument: TypeCoercion(
                                TypeCoercion {
                                    argument: Boolean(
                                        Boolean {
                                            value: false,
                                            source_information: SourceInformation {
                                                source_name: "",
                                                location: Location {
                                                    line_number: 0,
                                                    column_number: 0,
                                                },
                                                line: "",
                                            },
                                        },
                                    ),
                                    from: Boolean(
                                        Boolean {
                                            source_information: SourceInformation {
                                                source_name: "",
                                                location: Location {
                                                    line_number: 0,
                                                    column_number: 0,
                                                },
                                                line: "",
                                            },
                                        },
                                    ),
                                    to: Union(
                                        Union {
                                            types: {
                                                Boolean(
                                                    Boolean {
                                                        source_information: SourceInformation {
                                                            source_name: "",
                                                            location: Location {
                                                                line_number: 0,
                                                                column_number: 0,
                                                            },
                                                            line: "",
                                                        },
                                                    },
                                                ),
                                                None(
                                                    None {
                                                        source_information: SourceInformation {
                                                            source_name: "",
                                                            location: Location {
                                                                line_number: 0,
                                                                column_number: 0,
                                                            },
                                                            line: "",
                                                        },
                                                    },
                                                ),
                                            },
                                            source_information: SourceInformation {
                                                source_name: "",
                                                location: Location {
                                                    line_number: 0,
                                                    column_number: 0,
                                                },
                                                line: "",
                                            },
                                        },
                                    ),
                                    source_information: SourceInformation {
                                        source_name: "",
                                        location: Location {
                                            line_number: 0,
                                            column_number: 0,
                                        },
                                        line: "",
                                    },
                                },
                            ),
                            alternatives: [
                                Alternative {
                                    type_: Boolean(
                                        Boolean {
                                            source_information: SourceInformation {
                                                source_name: "",
                                                location: Location {
                                                    line_number: 0,
                                                    column_number: 0,
                                                },
                                                line: "",
                                            },
                                        },
                                    ),
                                    expression: TypeCoercion(
                                        TypeCoercion {
                                            argument: Number(
                                                Number {
                                                    value: 42.0,
                                                    source_information: SourceInformation {
                                                        source_name: "",
                                                        location: Location {
                                                            line_number: 0,
                                                            column_number: 0,
                                                        },
                                                        line: "",
                                                    },
                                                },
                                            ),
                                            from: Number(
                                                Number {
                                                    source_information: SourceInformation {
                                                        source_name: "",
                                                        location: Location {
                                                            line_number: 0,
                                                            column_number: 0,
                                                        },
                                                        line: "",
                                                    },
                                                },
                                            ),
                                            to: Union(
                                                Union {
                                                    types: {
                                                        None(
                                                            None {
                                                                source_information: SourceInformation {
                                                                    source_name: "",
                                                                    location: Location {
                                                                        line_number: 0,
                                                                        column_number: 0,
                                                                    },
                                                                    line: "",
                                                                },
                                                            },
                                                        ),
                                                        Number(
                                                            Number {
                                                                source_information: SourceInformation {
                                                                    source_name: "",
                                                                    location: Location {
                                                                        line_number: 0,
                                                                        column_number: 0,
                                                                    },
                                                                    line: "",
                                                                },
                                                            },
                                                        ),
                                                    },
                                                    source_information: SourceInformation {
                                                        source_name: "",
                                                        location: Location {
                                                            line_number: 0,
                                                            column_number: 0,
                                                        },
                                                        line: "",
                                                    },
                                                },
                                            ),
                                            source_information: SourceInformation {
                                                source_name: "",
                                                location: Location {
                                                    line_number: 0,
                                                    column_number: 0,
                                                },
                                                line: "",
                                            },
                                        },
                                    ),
                                },
                                Alternative {
                                    type_: None(
                                        None {
                                            source_information: SourceInformation {
                                                source_name: "",
                                                location: Location {
                                                    line_number: 0,
                                                    column_number: 0,
                                                },
                                                line: "",
                                            },
                                        },
                                    ),
                                    expression: TypeCoercion(
                                        TypeCoercion {
                                            argument: None(
                                                None {
                                                    source_information: SourceInformation {
                                                        source_name: "",
                                                        location: Location {
                                                            line_number: 0,
                                                            column_number: 0,
                                                        },
                                                        line: "",
                                                    },
                                                },
                                            ),
                                            from: None(
                                                None {
                                                    source_information: SourceInformation {
                                                        source_name: "",
                                                        location: Location {
                                                            line_number: 0,
                                                            column_number: 0,
                                                        },
                                                        line: "",
                                                    },
                                                },
                                            ),
                                            to: Union(
                                                Union {
                                                    types: {
                                                        None(
                                                            None {
                                                                source_information: SourceInformation {
                                                                    source_name: "",
                                                                    location: Location {
                                                                        line_number: 0,
                                                                        column_number: 0,
                                                                    },
                                                                    line: "",
                                                                },
                                                            },
                                                        ),
                                                        Number(
                                                            Number {
                                                                source_information: SourceInformation {
                                                                    source_name: "",
                                                                    location: Location {
                                                                        line_number: 0,
                                                                        column_number: 0,
                                                                    },
                                                                    line: "",
                                                                },
                                                            },
                                                        ),
                                                    },
                                                    source_information: SourceInformation {
                                                        source_name: "",
                                                        location: Location {
                                                            line_number: 0,
                                                            column_number: 0,
                                                        },
                                                        line: "",
                                                    },
                                                },
                                            ),
                                            source_information: SourceInformation {
                                                source_name: "",
                                                location: Location {
                                                    line_number: 0,
                                                    column_number: 0,
                                                },
                                                line: "",
                                            },
                                        },
                                    ),
                                },
                            ],
                            source_information: SourceInformation {
                                source_name: "",
                                location: Location {
                                    line_number: 0,
                                    column_number: 0,
                                },
                                line: "",
                            },
                        },
                    ),
                    type_: Union(
                        Union {
                            types: {
                                None(
                                    None {
                                        source_information: SourceInformation {
                                            source_name: "",
                                            location: Location {
                                                line_number: 0,
                                                column_number: 0,
                                            },
                                            line: "",
                                        },
                                    },
                                ),
                                Number(
                                    Number {
                                        source_information: SourceInformation {
                                            source_name: "",
                                            location: Location {
                                                line_number: 0,
                                                column_number: 0,
                                            },
                                            line: "",
                                        },
                                    },
                                ),
                            },
                            source_information: SourceInformation {
                                source_name: "",
                                location: Location {
                                    line_number: 0,
                                    column_number: 0,
                                },
                                line: "",
                            },
                        },
                    ),
                    source_information: SourceInformation {
                        source_name: "",
                        location: Location {
                            line_number: 0,
                            column_number: 0,
                        },
                        line: "",
                    },
                },
            ),
        ],
        export: Export {
            names: {},
        },
        imports: [],
    },
)
